uproot/source/compressed.py

```diff
@@ -127,8 +130,17 @@ def _prepare(self):
                     compression = self.compression.copy(uproot.const.kLZMA)
                 elif algo == b"L4":
                     compression = self.compression.copy(uproot.const.kLZ4)
-                    cursor.skip(8)        # FIXME: use this checksum!
                     compressedbytes -= 8
+                    checksum = cursor.bytes(self._compressed, 8).tolist()
+                    copy_cursor = copy(cursor)
+                    after_compressed = copy_cursor.bytes(self._compressed, compressedbytes)
+                    if sys.version_info < (3,):
+                        def iterbytes(data):
+                            return (ord (char) for char in data)
+                    else:
+                        iterbytes = lambda x: iter(x)
+                    if [x for x in iterbytes(xxhash.xxh64(after_compressed).digest())] != checksum:
```

Comment:
Does this iterate over all (compressed) bytes in the basket with a Python for loop? Maybe that's not an error but it would be a horrendous performance bug (on the compression method chosen for speed). Same for line 134 above: `numpy.ndarray.tolist` and `for x in numpy.ndarray` are not allowed (same for writing, when the dataset is large enough).

`xxhash` must have an interface that takes a whole array or a whole bytestring or a whole memoryview at a time. The number of steps in Python must not scale with the size of the data to compress or uncompress.